File: src/rubiks_cube/engine/move_definitions.py
from dataclasses import dataclass
from typing import Tuple


@dataclass(frozen=True)
class Move:
    notation: str
    axis: str  # 'x', 'y', 'z'
    layer: int  # -1, 0, 1
    direction: int  # 1 or -1


def normal_to_axis_layer_direction(normal, direction=1) -> Move:
    """Convert a normal vector (tuple or Vec3-like) into a Move.

    normal: object with .x/.y/.z or a 3-tuple
    """
    try:
        x, y, z = int(normal.x), int(normal.y), int(normal.z)
    except Exception:
        x, y, z = map(int, normal)

    if x != 0:
        return Move(
            notation=f"R" if x == 1 else "L", axis="x", layer=x, direction=direction
        )
    if y != 0:
        return Move(
            notation=f"U" if y == 1 else "D", axis="y", layer=y, direction=direction
        )
    if z != 0:
        return Move(
            notation=f"B" if z == 1 else "F", axis="z", layer=z, direction=direction
        )

    raise ValueError("zero normal is not a valid face selector")

File: frontend/src/engine/move_definitions.ts
export type Move = {
  notation: string
  axis: 'x' | 'y' | 'z'
  layer: number // -1, 0, 1 (we use 0..2 grid indices but keep semantics)
  direction: 1 | -1
}

export function normalToMove(normal: { x?: number; y?: number; z?: number } | [number, number, number], direction: 1 | -1 = 1): Move {
  let x: number, y: number, z: number
  if (Array.isArray(normal)) {
    ;[x, y, z] = normal.map((n) => Number(n)) as [number, number, number]
  } else {
    x = Number((normal as any).x || 0)
    y = Number((normal as any).y || 0)
    z = Number((normal as any).z || 0)
  }

  if (x !== 0) {
    return {
      notation: x === 1 ? 'R' : 'L',
      axis: 'x',
      layer: x,
      direction,
    }
  }
  if (y !== 0) {
    return {
      notation: y === 1 ? 'U' : 'D',
      axis: 'y',
      layer: y,
      direction,
    }
  }
  if (z !== 0) {
    return {
      notation: z === 1 ? 'B' : 'F',
      axis: 'z',
      layer: z,
      direction,
    }
  }

  throw new Error('zero normal is not a valid face selector')
}

export default { normalToMove }
